File: app/treasure.py
# -*- coding: utf-8 -*-
"""Tesouro do grupo: itens e moedas da mesa, separados das fichas.

Nada aqui mexe no inventário ou nas moedas pessoais. "Dividir" e "entregar"
tiram do tesouro e registram quem recebeu quanto; cada jogador anota na própria
ficha. Assim ninguém tem a ficha alterada sem ver.

Campaign.treasure (JSON):
    items  [{id, name, qty, weight, value, note}]
    coins  {chave_da_moeda: quantidade}
    log    [{at, who, text}]   — as 40 últimas movimentações
"""
import secrets
from datetime import datetime

from app.utils import to_float, to_int

MAX_ITEMS = 300
MAX_LOG = 40
MAX_AMOUNT = 10 ** 12


class TreasureError(ValueError):
    pass


def currencies(system):
    coins = [c for c in (system.data or {}).get("currencies") or [] if c.get("key")]
    return coins or [{"key": "moedas", "name": "Moedas", "abbr": ""}]


def normalize(raw):
    raw = raw if isinstance(raw, dict) else {}
    items = []
    for item in raw.get("items") or []:
        if not isinstance(item, dict) or not str(item.get("name") or "").strip():
            continue
        items.append({
            "id": str(item.get("id") or secrets.token_hex(4))[:16],
            "name": str(item["name"]).strip()[:120],
            "qty": max(0, min(MAX_AMOUNT, to_int(item.get("qty"), 1))),
            "weight": max(0.0, to_float(item.get("weight"), 0.0)),
            "value": str(item.get("value") or "").strip()[:60],
            "note": str(item.get("note") or "").strip()[:200],
        })
    coins = {}
    for key, amount in (raw.get("coins") or {}).items():
        coins[str(key)[:40]] = max(0, min(MAX_AMOUNT, to_int(amount, 0)))
    log = [entry for entry in raw.get("log") or [] if isinstance(entry, dict)][:MAX_LOG]
    return {"items": items[:MAX_ITEMS], "coins": coins, "log": log}


def _log(treasure, who, text):
    treasure["log"].insert(0, {"at": datetime.utcnow().isoformat() + "Z", "who": who,
                               "text": text[:300]})
    del treasure["log"][MAX_LOG:]


def _item(treasure, item_id):
    for item in treasure["items"]:
        if item["id"] == item_id:
            return item
    raise TreasureError("Esse item não está mais no tesouro.")


def _coin_label(coins, key):
    for coin in coins:
        if coin["key"] == key:
            return coin.get("abbr") or coin.get("name") or key
    return key


def apply(treasure, payload, who, coins, party):
    """Uma operação vinda da tela. `coins`: moedas do sistema; `party`: {id: nome} dos PJs."""
    op = payload.get("op")
    keys = {c["key"] for c in coins}

    if op == "add_item":
        name = str(payload.get("name") or "").strip()[:120]
        if not name:
            raise TreasureError("Dê um nome ao item.")
        if len(treasure["items"]) >= MAX_ITEMS:
            raise TreasureError("O tesouro já tem itens demais.")
        qty = max(1, min(MAX_AMOUNT, to_int(payload.get("qty"), 1)))
        treasure["items"].append({
            "id": secrets.token_hex(4), "name": name, "qty": qty,
            "weight": max(0.0, to_float(payload.get("weight"), 0.0)),
            "value": str(payload.get("value") or "").strip()[:60],
            "note": str(payload.get("note") or "").strip()[:200],
        })
        _log(treasure, who, "guardou %d× %s" % (qty, name))

    elif op == "update_item":
        item = _item(treasure, str(payload.get("id")))
        if "name" in payload and str(payload["name"]).strip():
            item["name"] = str(payload["name"]).strip()[:120]
        if "qty" in payload:
            item["qty"] = max(0, min(MAX_AMOUNT, to_int(payload.get("qty"), item["qty"])))
        if "weight" in payload:
            item["weight"] = max(0.0, to_float(payload.get("weight"), item["weight"]))
        for key, limit in (("value", 60), ("note", 200)):
            if key in payload:
                item[key] = str(payload.get(key) or "").strip()[:limit]

    elif op == "remove_item":
        item = _item(treasure, str(payload.get("id")))
        treasure["items"].remove(item)
        _log(treasure, who, "tirou %s do tesouro" % item["name"])

    elif op == "give_item":
        item = _item(treasure, str(payload.get("id")))
        target = party.get(to_int(payload.get("to"), 0))
        if target is None:
            raise TreasureError("Escolha um personagem do grupo.")
        qty = max(1, to_int(payload.get("qty"), 1))
        if qty > item["qty"]:
            raise TreasureError("Só há %d× %s no tesouro." % (item["qty"], item["name"]))
        item["qty"] -= qty
        if item["qty"] == 0:
            treasure["items"].remove(item)
        _log(treasure, who, "entregou %d× %s a %s" % (qty, item["name"], target))

    elif op == "coins":
        deltas = payload.get("deltas") or {}
        changes = []
        for key, raw in deltas.items():
            if key not in keys:
                continue
            delta = to_int(raw, 0)
            if not delta:
                continue
            current = treasure["coins"].get(key, 0)
            if current + delta < 0:
                raise TreasureError("Não há %s suficiente: o tesouro tem %d."
                                    % (_coin_label(coins, key), current))
            treasure["coins"][key] = min(MAX_AMOUNT, current + delta)
            changes.append("%s%d %s" % ("+" if delta > 0 else "", delta, _coin_label(coins, key)))
        if not changes:
            raise TreasureError("Nada para mudar.")
        reason = str(payload.get("reason") or "").strip()[:120]
        _log(treasure, who, ", ".join(changes) + (" — " + reason if reason else ""))

    elif op == "split":
        chosen = [cid for cid in (to_int(x, 0) for x in payload.get("characters") or [])
                  if cid in party]
        chosen = list(dict.fromkeys(chosen))
        if not chosen:
            raise TreasureError("Escolha quem entra na divisão.")
        wanted = [k for k in (payload.get("coins") or list(keys)) if k in keys]
        shares = split_preview(treasure, coins, len(chosen), wanted)
        if not any(s["each"] for s in shares):
            raise TreasureError("Não há moedas suficientes para dividir entre %d." % len(chosen))
        parts = []
        for share in shares:
            if not share["each"]:
                continue
            treasure["coins"][share["key"]] -= share["each"] * len(chosen)
            parts.append("%d %s" % (share["each"], share["label"]))
        leftover = ["%d %s" % (s["left"], s["label"]) for s in shares if s["left"]]
        names = ", ".join(party[cid] for cid in chosen)
        _log(treasure, who, "dividiu entre %s: %s para cada um%s"
             % (names, " + ".join(parts),
                " (sobrou " + ", ".join(leftover) + " no tesouro)" if leftover else ""))
    else:
        raise TreasureError("Operação desconhecida.")
    return treasure


def split_preview(treasure, coins, people, keys=None):
    """Quanto cada um recebe e quanto sobra, por moeda."""
    out = []
    for coin in coins:
        if keys is not None and coin["key"] not in keys:
            continue
        total = treasure["coins"].get(coin["key"], 0)
        each = total // people if people else 0
        out.append({"key": coin["key"], "label": coin.get("abbr") or coin.get("name"),
                    "each": each, "left": total - each * people})
    return out


def total_weight(treasure):
    return round(sum(i["qty"] * i["weight"] for i in treasure["items"]), 2)

File: app/static/js/treasure.js
/* Tesouro do grupo: moedas, itens, divisão e registro.

   Qualquer um da mesa mexe; o servidor aplica cada operação sobre o valor mais
   novo (não perde a mudança de outra pessoa) e devolve o tesouro atualizado.
   As mudanças dos outros chegam pela consulta única (window.Live).
*/
(function () {
  "use strict";

  var root = document.querySelector("[data-treasure]");
  if (!root) return;

  var state;
  try { state = JSON.parse(root.querySelector("[data-treasure-payload]").textContent); } catch (e) { return; }
  var status = root.querySelector("[data-treasure-status]");
  var chosen = {};  // quem entra na divisão (id -> true)
  state.party.forEach(function (p) { chosen[p.id] = true; });

  function el(tag, className, text) {
    var node = document.createElement(tag);
    if (className) node.className = className;
    if (text !== undefined) node.textContent = text;
    return node;
  }

  function coinLabel(coin) { return coin.abbr || coin.name; }
  function number(value) { return Number(String(value || "").replace(",", ".")) || 0; }
  function pretty(value) {
    var n = Math.round(value * 100) / 100;
    return String(n).replace(".", ",");
  }

  function send(body) {
    status.textContent = "salvando…";
    status.dataset.kind = "saving";
    return window.api(root.dataset.url, { body: body }).then(function (data) {
      if (!data.ok) {
        status.textContent = data.message || "não deu certo";
        status.dataset.kind = "error";
        return false;
      }
      status.textContent = "salvo ✓";
      status.dataset.kind = "saved";
      adopt(data.treasure);
      if (window.Live) window.Live.poke();
      return true;
    }).catch(function () {
      status.textContent = "sem conexão";
      status.dataset.kind = "error";
      return false;
    });
  }

  /* ------------------------------------------------------------ moedas */
  function renderCoins() {
    var grid = root.querySelector("[data-coins]");
    grid.innerHTML = "";
    state.coins.forEach(function (coin) {
      var box = el("div", "coin");
      box.appendChild(el("span", "coin-amount", String(coin.amount)));
      box.appendChild(el("span", "coin-name", coinLabel(coin)));
      box.title = coin.name;
      grid.appendChild(box);
    });
    var inputs = root.querySelector("[data-coin-inputs]");
    if (!inputs.children.length) {
      state.coins.forEach(function (coin) {
        var label = el("label", "coin-input");
        var input = el("input");
        input.type = "number";
        input.name = "coin_" + coin.key;
        input.dataset.coin = coin.key;
        input.placeholder = "0";
        input.setAttribute("aria-label", coin.name);
        label.appendChild(input);
        label.appendChild(el("span", "muted small", coinLabel(coin)));
        inputs.appendChild(label);
      });
    }
  }

  root.querySelector("[data-coin-form]").addEventListener("submit", function (event) {
    event.preventDefault();
    var form = event.target, deltas = {}, any = false;
    form.querySelectorAll("[data-coin]").forEach(function (input) {
      var value = parseInt(input.value, 10);
      if (value) { deltas[input.dataset.coin] = value; any = true; }
    });
    if (!any) return;
    var reason = form.elements["reason"];
    send({ op: "coins", deltas: deltas, reason: reason.value }).then(function (ok) {
      if (!ok) return;
      form.querySelectorAll("[data-coin]").forEach(function (input) { input.value = ""; });
      reason.value = "";
    });
  });

  /* ------------------------------------------------------------ divisão */
  function renderSplit() {
    var box = root.querySelector("[data-split-people]");
    box.innerHTML = "";
    if (!state.party.length) {
      box.appendChild(el("p", "muted small", "Nenhum personagem de jogador na campanha ainda."));
    }
    state.party.forEach(function (person) {
      var label = el("label", "check");
      var check = el("input");
      check.type = "checkbox";
      check.checked = !!chosen[person.id];
      check.addEventListener("change", function () { chosen[person.id] = check.checked; renderPreview(); });
      label.appendChild(check);
      label.appendChild(document.createTextNode(" " + person.name));
      box.appendChild(label);
    });
    renderPreview();
  }

  function people() {
    return state.party.filter(function (p) { return chosen[p.id]; });
  }

  function renderPreview() {
    var preview = root.querySelector("[data-split-preview]");
    var n = people().length;
    if (!n) { preview.textContent = "Marque quem entra na divisão."; return; }
    var parts = state.coins.filter(function (c) { return c.amount; }).map(function (coin) {
      var each = Math.floor(coin.amount / n), left = coin.amount - each * n;
      return each + " " + coinLabel(coin) + (left ? " (sobra " + left + ")" : "");
    });
    preview.textContent = parts.length ? "Cada um recebe: " + parts.join(" · ") : "O tesouro está sem moedas.";
  }

  root.querySelector("[data-split]").addEventListener("click", function () {
    var list = people();
    if (!list.length) return;
    var names = list.map(function (p) { return p.name; }).join(", ");
    if (!confirm("Dividir as moedas entre " + names + "? As partes saem do tesouro; cada um anota a sua na ficha.")) return;
    send({ op: "split", characters: list.map(function (p) { return p.id; }) });
  });

  /* ------------------------------------------------------------- itens */
  function renderItems() {
    var box = root.querySelector("[data-items]");
    box.innerHTML = "";
    root.querySelector("[data-weight]").textContent = state.weight
      ? "peso total " + pretty(state.weight) + (state.unit ? " " + state.unit : "") : "";
    if (!state.items.length) {
      box.appendChild(el("p", "muted small", "Nada guardado ainda."));
      return;
    }
    state.items.forEach(function (item) {
      var row = el("div", "treasure-item");
      var main = el("div", "treasure-item-main");
      main.appendChild(el("strong", "", item.name));
      var meta = [];
      if (item.weight) meta.push(pretty(item.weight) + (state.unit ? " " + state.unit : "") + " cada");
      if (item.value) meta.push(item.value);
      if (item.note) meta.push(item.note);
      if (meta.length) main.appendChild(el("span", "muted small", meta.join(" · ")));
      row.appendChild(main);

      var qty = el("input");
      qty.type = "number";
      qty.min = "0";
      qty.value = item.qty;
      qty.className = "treasure-qty";
      qty.setAttribute("aria-label", "Quantidade de " + item.name);
      qty.addEventListener("change", function () {
        send({ op: "update_item", id: item.id, qty: qty.value });
      });
      row.appendChild(qty);

      if (state.party.length) {
        var give = el("select", "treasure-give");
        give.setAttribute("aria-label", "Entregar " + item.name);
        var first = el("option", "", "Entregar a…");
        first.value = "";
        give.appendChild(first);
        state.party.forEach(function (person) {
          var option = el("option", "", person.name);
          option.value = person.id;
          give.appendChild(option);
        });
        give.addEventListener("change", function () {
          if (!give.value) return;
          var person = state.party.filter(function (p) { return String(p.id) === give.value; })[0];
          var how = item.qty > 1 ? prompt("Quantos " + item.name + " para " + person.name + "?", "1") : "1";
          if (how === null) { give.value = ""; return; }
          send({ op: "give_item", id: item.id, to: person.id, qty: parseInt(how, 10) || 1 });
        });
        row.appendChild(give);
      }

      var remove = el("button", "btn btn-ghost btn-sm btn-icon", "✕");
      remove.type = "button";
      remove.title = "Tirar do tesouro";
      remove.addEventListener("click", function () {
        if (confirm("Tirar “" + item.name + "” do tesouro?")) send({ op: "remove_item", id: item.id });
      });
      row.appendChild(remove);
      box.appendChild(row);
    });
  }

  root.querySelector("[data-item-form]").addEventListener("submit", function (event) {
    event.preventDefault();
    var f = event.target.elements;
    send({ op: "add_item", name: f["name"].value, qty: f["qty"].value,
           weight: number(f["weight"].value), value: f["value"].value }).then(function (ok) {
      if (!ok) return;
      f["name"].value = "";
      f["qty"].value = "1";
      f["weight"].value = "";
      f["value"].value = "";
      f["name"].focus();
    });
  });

  /* ---------------------------------------------------------- registro */
  function renderLog() {
    var box = root.querySelector("[data-log]");
    box.innerHTML = "";
    if (!state.log.length) {
      box.appendChild(el("p", "muted small", "Nenhuma movimentação ainda."));
      return;
    }
    state.log.forEach(function (entry) {
      var line = el("div", "treasure-log-line");
      var when = "";
      var parsed = new Date(entry.at);
      if (!isNaN(parsed)) {
        when = parsed.toLocaleString("pt-BR", { day: "2-digit", month: "2-digit",
                                                hour: "2-digit", minute: "2-digit" });
      }
      line.appendChild(el("span", "muted small", when));
      line.appendChild(el("strong", "", entry.who));
      line.appendChild(el("span", "", entry.text));
      box.appendChild(line);
    });
  }

  function adopt(data) {
    state = data;
    state.party.forEach(function (p) { if (chosen[p.id] === undefined) chosen[p.id] = true; });
    renderCoins();
    renderSplit();
    renderItems();
    renderLog();
  }

  if (window.Live && window.Live.enabled) {
    var handle = window.Live.register("treasure", "", function (data) {
      // Não redesenha embaixo de quem está digitando uma quantidade.
      if (root.contains(document.activeElement) && document.activeElement.tagName !== "BUTTON") {
        handle.reset();
        return;
      }
      adopt(data);
    });
  }
  adopt(state);
})();
